File: data_worker/src/iface/common/grant_state_manager.py
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Literal

from src.models.auth.forwarded_auth import ForwardedAuthContext


GrantStatus = Literal["active", "used", "revoked", "expired"]


@dataclass(slots=True, kw_only=True)
class GrantStateRecord:
    grant_key: str

    principal_id: str
    session_id: str
    token_id: str

    gateway_id: str = ""
    source_service: str = ""
    target_service: str = ""

    status: GrantStatus = "active"

    issued_at: float = 0.0
    expires_at: float = 0.0
    used_at: float = 0.0
    updated_at: float = 0.0

    metadata: dict[str, str] = field(default_factory=dict)


class IGrantStateManager(ABC):
    """可选的下游授权状态管理端口。"""

    @abstractmethod
    async def save_grant(self, grant: ForwardedAuthContext) -> str:
        raise NotImplementedError

    @abstractmethod
    async def load_active_grant(self, token_id: str) -> GrantStateRecord | None:
        raise NotImplementedError

    @abstractmethod
    async def mark_grant_used(self, token_id: str, used_at: float) -> None:
        raise NotImplementedError

    @abstractmethod
    async def revoke_grant(self, token_id: str, reason: str = "") -> None:
        raise NotImplementedError

    @abstractmethod
    async def purge_expired_grants(self, before: float) -> int:
        raise NotImplementedError

File: certification_server/src/iface/common/grant_state_manager.go
package common

import (
	"context"
	"time"

	authmodel "certification_server/src/models/auth"

	"github.com/google/uuid"
)

// GrantStatus 定义下游授权令牌的生命周期状态。
type GrantStatus string

const (
	GrantStatusActive  GrantStatus = "active"
	GrantStatusUsed    GrantStatus = "used"
	GrantStatusRevoked GrantStatus = "revoked"
	GrantStatusExpired GrantStatus = "expired"
)

// GrantStateRecord 是下游授权令牌的持久化快照。
type GrantStateRecord struct {
	GrantKey string

	SessionID   uuid.UUID
	TokenID     uuid.UUID
	PrincipalID string

	GatewayID     string
	SourceService string
	TargetService string

	Status GrantStatus

	IssuedAt  time.Time
	ExpiresAt time.Time
	UsedAt    time.Time
	UpdatedAt time.Time

	Metadata map[string]string
}

// IGrantStateManager 定义可选的下游授权状态管理。
type IGrantStateManager interface {
	SaveGrant(ctx context.Context, grant *authmodel.DownstreamAccessGrant) (string, error)
	LoadActiveGrant(ctx context.Context, tokenID uuid.UUID) (*GrantStateRecord, error)
	MarkGrantUsed(ctx context.Context, tokenID uuid.UUID, usedAt time.Time) error
	RevokeGrant(ctx context.Context, tokenID uuid.UUID, reason string) error
	PurgeExpiredGrants(ctx context.Context, before time.Time) (int64, error)
}
